### PR TITLE
PLFM-5842: Added quarantine email management

### PR DESCRIPTION
**Pull request by marcomarasca**

### CODE REVIEW

File: lib/jdomodels/src/main/resources/schema/ses/SESNotifications-ddl.sql [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
+  `INSTANCE_NUMBER` INT NOT NULL,
   `SES_MESSAGE_ID` VARCHAR(256) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
   `SES_FEEDBACK_ID` VARCHAR(256) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
   `NOTIFICATION_TYPE` ENUM('BOUNCE', 'COMPLAINT', 'DELIVERY', 'UNKNOWN') NOT NULL,
-  `NOTIFICATION_SUBTYPE` VARCHAR(20) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
-  `NOTIFICATION_REASON` VARCHAR(20) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
-  `NOTIFICATION_BODY` MEDIUMBLOB,
+  `NOTIFICATION_SUBTYPE` VARCHAR(256) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
+  `NOTIFICATION_REASON` VARCHAR(256) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
+  `NOTIFICATION_BODY` MEDIUMBLOB NOT NULL,
```
Comment (zimingd) [inline lib/jdomodels/src/main/resources/schema/ses/SESNotifications-ddl.sql]: Set type to JSON

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/ses/SESNotification.java [hunk 2]
```diff
@@ -12,6 +14,7 @@
 public class SESNotification {
```
Comment (zimingd) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/ses/SESNotification.java]: SESNotificationRecord

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/ses/EmailQuarantineDaoImplTest.java [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+		dao.clearAll();
+	}
+
+	@AfterEach
+	public void after() {
+		dao.clearAll();
+	}
+
+	@Test
+	public void testAddToQuarantineWithPastTimeout() {
```
Comment (zimingd) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/ses/EmailQuarantineDaoImplTest.java]: testAddToQuarantineWithInvalidtTimeoutValues

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/ses/EmailQuarantineDaoImplTest.java [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+	}
+
+	private QuarantinedEmail getTestQuarantinedEmail() {
+		return getTestQuarantinedEmail(testEmail, QuarantineReason.PERMANENT_BOUNCE);
+	}
+
+	private QuarantinedEmail getTestQuarantinedEmail(String email, QuarantineReason reason) {
+		return new QuarantinedEmail(email, reason).withSesMessageId(sesMessageId);
+	}
+
```
Comment (zimingd) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/ses/EmailQuarantineDaoImplTest.java:348]: add test with batch of multiple emails

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/ses/SESNotificationDaoImplTest.java [hunk 5]
```diff
@@ -19,127 +23,92 @@
...
+
+		Long result = dao.countBySesMessageId(sesMessageId);
+
 		assertEquals(1L, result);
 	}
-	
+
+	private List<SESNotification> getRandomNotifications(int count) {
+		List<SESNotification> list = new ArrayList<>(count);
+		IntStream.range(0, count).forEach(_index -> {
```
Comment (zimingd) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/ses/SESNotificationDaoImplTest.java]: for loop

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/ses/QuarantinedEmailBatch.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+ * @author Marco
+ *
+ */
+public class QuarantinedEmailBatch extends ArrayList<QuarantinedEmail> {
+
+	public static QuarantinedEmailBatch emptyBatch() {
+		return new QuarantinedEmailBatch();
+	}
+
+	private QuarantineReason reason;
```
Comment (zimingd) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/ses/QuarantinedEmailBatch.java:22]: There is currently a hidden nuance in behavior for adding quarantined emails:
String emails have the reason and messageId of what is set in QuarantinedEmailBatch, but adding a QuarantinedEmail object does not override its reason and messageId.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/ses/EmailQuarantineBounceProvider.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+			batch.withReason(QuarantineReason.TRANSIENT_BOUNCE).withExpirationTimeout(RETRY_TIMEOUT);
+		} else {
+			return batch;
+		}
+		
+		notificationDetails.getReason().ifPresent( reasonDetails -> {
+			batch.withReasonDetails(reasonDetails.toUpperCase());
+		});
+
+		notificationDetails.getRecipients().forEach(recipient -> {
```
Comment (zimingd) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/ses/EmailQuarantineBounceProvider.java]: use normal for each loop

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/ses/EmailQuarantineComplaintProvider.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	// https://docs.aws.amazon.com/ses/latest/DeveloperGuide/notification-contents.html#complaint-object)
+	private Set<String> COMPLAINT_REASONS = ImmutableSet.of("ABUSE", "AUTH-FAILURE", "FRAUD", "VIRUS", "OTHER");
+
+	@Override
+	public SESNotificationType getSupportedType() {
+		return SESNotificationType.COMPLAINT;
+	}
+
+	@Override
+	public QuarantinedEmailBatch getQuarantinedEmails(SESJsonNotificationDetails notificationDetails, String sesMessageId) {
```
Comment (zimingd) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/ses/EmailQuarantineComplaintProvider.java:25]: Do not quarantine complaints. we should restrict the sender, not the recipient in this case.
For example:
fraud@fakecompany.com sends fake email to user@syanpse.org and that message gets marked as fraud. Under the current system user@syanpse.org would no longer receive emails even though this is a completely valid email address.

**Review by zimingd — CHANGES_REQUESTED** [review]

File: lib/jdomodels/src/main/resources/schema/ses/SESNotifications-ddl.sql (hunk 1 above)
Comment (marcomarasca) [reply lib/jdomodels/src/main/resources/schema/ses/SESNotifications-ddl.sql]: Changed to JSON

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/ses/SESNotification.java (hunk 2 above)
Comment (marcomarasca) [reply lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/ses/SESNotification.java]: Renamed

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/ses/EmailQuarantineDaoImplTest.java (hunk 3 above)
Comment (marcomarasca) [reply lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/ses/EmailQuarantineDaoImplTest.java]: Renamed

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/ses/EmailQuarantineDaoImplTest.java (hunk 4 above)
Comment (marcomarasca) [reply lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/ses/EmailQuarantineDaoImplTest.java:348]: Will do in the next PR

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/ses/SESNotificationDaoImplTest.java (hunk 5 above)
Comment (marcomarasca) [reply lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/ses/SESNotificationDaoImplTest.java]: Done

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/ses/QuarantinedEmailBatch.java (hunk 6 above)
Comment (marcomarasca) [reply lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/ses/QuarantinedEmailBatch.java:22]: Will rethink in next PR

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/ses/EmailQuarantineBounceProvider.java (hunk 7 above)
Comment (marcomarasca) [reply services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/ses/EmailQuarantineBounceProvider.java]: Done

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/ses/EmailQuarantineComplaintProvider.java (hunk 8 above)
Comment (marcomarasca) [reply services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/ses/EmailQuarantineComplaintProvider.java:25]: I removed this for now, but from the data sent from SES it does not appear to be the case. Looks like the recipients are the offending ones, we will have to monitor this.

**Review by zimingd — APPROVED** [review]

**Pull request merged**